src/box/lua/net_box.c

```diff
@@ -1764,13 +1778,15 @@ luaT_netbox_request_pairs(struct lua_State *L)
 
 /**
  * Creates a netbox transport object (userdata) and pushes it to Lua stack.
+ * Takes the on_event callback.
  */
 static int
 netbox_new_transport(struct lua_State *L)
 {
+	int on_event_ref = luaL_ref(L, LUA_REGISTRYINDEX);
 	struct netbox_transport *transport = lua_newuserdata(
 		L, sizeof(*transport));
-	netbox_transport_create(transport);
+	netbox_transport_create(transport, on_event_ref);
```

Comment:
Why do you need this on the level of watcher API? You can do both mentioned things on top of it. 

> With a remote watcher, we need to send an ack to the server the first time a value is processed.

You will have `watcher_run_f` which calls a Lua function. Here you can send the ACK after the Lua function is finished.

> We also need to send the watch command the first time a watcher is registered for a key and the unwatch command when the last watcher is unregistered.

Is it really that hard? You can expose a function like `watchable_has_key()` and if it returns `false`, then after `watchable_register_watcher` you send the `WATCH` command.
The same for unregistration. Before the watcher is deleted in `watcher_destroy_f`, you check if now `watchable_has_key()` for its key return false. If it does - send `UNWATCH`.

> making watchable_node public

It is already public, you put it into the header. But are the functions you need? Above I could only find `watchable_has_key`, however it is not a method of `watchable_node`.

Anyway, if you don't want, then lets deal with Lua.

Reply:
> You will have watcher_run_f which calls a Lua function. Here you can send the ACK after the Lua function is finished.

The problem is there may be more than one watcher registered for the same key, in which case we still want to send exactly one ACK per value. If we simply send an ACK whenever a watcher Lua function returns, we will end up sending an ACK per each registered watcher for each received value. We'd have to propagate some extra information from the watcher lib to the net.box callback to avoid that - take a look at how `watcher:_run()` handles it.

> Is it really that hard? You can expose a function like watchable_has_key() and if it returns false, then after watchable_register_watcher you send the WATCH command.

When registering the first net.box watcher for a key, we don't want to notify it until we receive the key value from the server. It isn't the case with a local watcher, which is run unconditionally on registration, so if we did what you suggested, we'd invoke net.box watchers with an invalid value (null) after registration. Note the difference between `remote_methods:watch()` and `watchable_register_watcher()`.

> Anyway, if you don't want, then lets deal with Lua.

It's not that I don't want to - actually, that's what I intended to do initially, but as I said the resulting API turned out to be quite complicated and confusing. Since the watcher library is trivial and net.box is an isolated module, I reckoned it should be fine to re-implement watchers for net.box without reusing the local watcher code.

Anyway, if you or anyone else manages to reuse the local watchers code in net.box so that the resulting watcher API doesn't look ugly, I'll be happy.